wefe/debias/repulsion_attraction_neutralization.py

```diff
@@ -0,0 +1,495 @@
+from copy import deepcopy
+from typing import Dict, Any, Optional, List, Sequence
+
+from tqdm import tqdm
+from wefe.debias.base_debias import BaseDebias
+from sklearn.decomposition import PCA
+from wefe.preprocessing import EmbeddingDict, get_embeddings_from_sets
+import numpy as np
+from wefe.utils import check_is_fitted
+from wefe.word_embedding_model import WordEmbeddingModel
+import torch.nn as nn
+import torch
+from copy import deepcopy
+
+
+class RepulsionAttractionNeutralization(BaseDebias):
+    """Repulsion attraction Neutralization method.
+
+    This method allow reducing the bias of an embedding model creating a transformation
+    such that the stereotypical gender information are minimized with minimal semantic offset
+
+    This method is binary because it only allows 2 classes of the same bias criterion,
+    such as male or female.
+    For a multiclass debias (such as for Latinos, Asians and Whites), it is recommended
+    to visit MulticlassHardDebias class.
+
+    The main idea of this method is:
+
+    1. **Identify a bias subspace through the defining sets.** In the case of gender,
+    these could be e.g. `{'woman', 'man'}, {'she', 'he'}, ...`
+
+    2. A multiobjective optimization is performed. For each vector w in the target set
+    it is found its debias counterpart wd by solving:
+
+    argmin(Fr(wd),Fa(wd),Fn(wd))
+
+    where Fr, Fa, Fn are repulsion, attraction and neutralization functions defined as the following:
+
+    Fr(wd) =  Σ |cos(wd,ni)| / |S|
+    Repulsion repels w from the repulsion (S) set defined as the n words closer to w whose indirect bias is grater thar a threshold.
+
+    Fa(wd) = |cos(wd,w)-1|/2
+    Attraction atracts wd to the original vector w to minimize the loss of semantic
+
+    Fn(wd) = |cos(wd,g)|
+    Neutralization minimize the bias to a particular gender. With g being de bias direction.
+
+    The optimization is performed by formulating a single objective:
+    F(wd) =  λ1Fr(wd) + λ2Fa(wd) + λ3Fn(wd)
+
+
```

Comment:
Here I would also add the equivalence between the paper's definitions of Preserve and Debias set and WEFE's target and ignore.

Reply:
both sets included in the documentation